### PR TITLE
Tempus: app action for newmark implicit a form

### LINKED ISSUE #7893: Tempus: AppAction for Newmark Implicit A Form
## Enhancement
@trilinos/tempus 

Add AppAction interface for Newmark Implicit A Form stepper.

### PR DESCRIPTION
**Pull request by sconde**
@trilinos/tempus

### CODE REVIEW

File: packages/tempus/src/Tempus_StepperNewmarkImplicitAFormModifierXBase.hpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ *  \renewcommand{\thealgorithm}{}
+ *  \caption{Newmark Implicit-A with modify calls indicated.}
+ *  \begin{algorithmic}[1]
+ *    \State \quad {\it modifierX.modify(x, time, dt, X\_BEGIN\_STEP)}
+ *    \State Compute the predictor (e.g., apply stepper to $x_n$).
+ *    \State \quad {\it modifierX.modify(x, time, dt, X\_BEFORE\_SOLVE)}
+ *    \State Solve $\mathcal{F}_n(\dot{x}=(x_n-x_{n-1})/\Delta t_n, x_n, t_n)=0$ for $x_n$
+ *    \State \quad {\it modifierX.modify(x, time, dt, X\_AFTER\_SOLVE)}
+ *    \State $\dot{x}_n \leftarrow (x_n-x_{n-1})/\Delta t_n$
+ *    \State \quad {\it modifierX.modify(x, time, dt, XDOT\_END\_STEP)}
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperNewmarkImplicitAFormModifierXBase.hpp:44]: I think we just want X_END_STEP and not xDot.

File: packages/tempus/src/Tempus_StepperNewmarkImplicitAFormModifierXBase.hpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        x = workingState->getX();
+        break;
+      }
+      case StepperNewmarkImplicitAFormAppAction<Scalar>::END_STEP:
+      {
+        modType = XDOT_END_STEP;
+        if (workingState->getXDot() != Teuchos::null)
+          x = workingState->getXDot();
+        else
+          x = stepper->getStepperXDot();
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperNewmarkImplicitAFormModifierXBase.hpp:105]: Same here.

File: packages/tempus/src/Tempus_StepperNewmarkImplicitAFormModifierXBase.hpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  }
+
+public:
+
+  /// Indicates the location of application action (see algorithm).
+  enum MODIFIER_TYPE {
+    X_BEGIN_STEP,     ///< Modify \f$x\f$ at the beginning of the step.
+    X_BEFORE_SOLVE,   ///< Modify \f$x\f$ before the implicit solve.
+    X_AFTER_SOLVE,    ///< Modify \f$x\f$ after the implicit solve.
+    XDOT_END_STEP     ///< Modify \f$\dot{x}\f$ at the end of the step.
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperNewmarkImplicitAFormModifierXBase.hpp:123]: Same here.

File: packages/tempus/src/Tempus_StepperNewmarkImplicitAFormModifierXDefault.hpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  virtual void modify(
+    Teuchos::RCP<Thyra::VectorBase<Scalar> > /* x */,
+    const Scalar /* time */, const Scalar /* dt */,
+    const typename StepperNewmarkImplicitAFormModifierXBase<Scalar>::MODIFIER_TYPE modType)
+  {
+    switch(modType) {
+      case StepperNewmarkImplicitAFormModifierXBase<Scalar>::X_BEGIN_STEP:
+      case StepperNewmarkImplicitAFormModifierXBase<Scalar>::X_BEFORE_SOLVE:
+      case StepperNewmarkImplicitAFormModifierXBase<Scalar>::X_AFTER_SOLVE:
+      case StepperNewmarkImplicitAFormModifierXBase<Scalar>::XDOT_END_STEP:
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperNewmarkImplicitAFormModifierXDefault.hpp:48]: Same here.

File: packages/tempus/src/Tempus_StepperNewmarkImplicitAForm_impl.hpp [hunk 5]
```diff
@@ -207,8 +228,41 @@ StepperNewmarkImplicitAForm<Scalar>::StepperNewmarkImplicitAForm(
   this->setSchemeName(         schemeName);
   this->setBeta(               beta);
   this->setGamma(              gamma);
-
   this->setObserver(obs);
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperNewmarkImplicitAForm_impl.hpp:231]: Need TEMPUS_HIDE_DEPRECATED_CODE around the setObserver().

Comment (ccober6) [reply packages/tempus/src/Tempus_StepperNewmarkImplicitAForm_impl.hpp:231]: Oh, sorry.  You have it around the whole constructor.

**Review by ccober6 — CHANGES_REQUESTED** [review]
Looks good! Just a few changes.  Also can you squash your commits once you are done?  Thanks!

File: packages/tempus/src/Tempus_StepperNewmarkImplicitAFormModifierXDefault.hpp (hunk 4 above)
Comment (sconde) [reply packages/tempus/src/Tempus_StepperNewmarkImplicitAFormModifierXDefault.hpp:48]: Ok. I will take care of those immediately

**Comment by sconde** [comment]
@ccober6 I've squashed the commits. Can I get a review again please

**Review by ccober6 — APPROVED** [review]
Looks good! Thanks!

**Pull request merged**